src/sentry/api/endpoints/organization_member_team_details.py

```diff
@@ -150,6 +187,58 @@ def post(
 
         return Response(serialize(team, request.user, TeamWithProjectsSerializer()), status=201)
 
+    def put(
+        self,
+        request: Request,
+        organization: Organization,
+        member: OrganizationMember,
+        team_slug: str,
+    ) -> Response:
+        try:
+            team = Team.objects.get(organization=organization, slug=team_slug)
+        except Team.DoesNotExist:
+            raise ResourceDoesNotExist
+
+        try:
+            omt = OrganizationMemberTeam.objects.get(team=team, organizationmember=member)
+        except OrganizationMemberTeam.DoesNotExist:
+            raise ResourceDoesNotExist
+
+        serializer = OrganizationMemberTeamSerializer(data=request.data, partial=True)
+        if not serializer.is_valid():
+            return Response(status=400)
+        result = serializer.validated_data
+
+        if "role" in result and features.has("organizations:team-roles", organization):
+            new_role_id = result["role"]
+            try:
+                new_role = team_roles.get(new_role_id)
+            except KeyError:
+                return Response(status=400)
+
+            if not self._can_set_team_role(request, team, new_role):
+                return Response({"detail": ERR_INSUFFICIENT_ROLE}, status=400)
+
+            self._change_team_member_role(omt, new_role)
+
+        return Response(serialize(team, request.user, TeamWithProjectsSerializer()), status=200)
+
+    @staticmethod
+    def _change_team_member_role(member: OrganizationMemberTeam, role: TeamRole) -> None:
+        """Modify a member's team-level role.
+
+        If the member has an organization role that gives an equal or higher entry
+        role, write null to this object's role field. We do this because a persistent
+        team role, if it is overshadowed by the entry role, would be effectively
+        invisible in the UI, and would be surprising if it were left behind after the
+        user's org-level role is lowered.
+        """
+        entry_role = roles.get_entry_role(member.organizationmember.role)
+        if role.priority > entry_role.priority:
+            member.update(role=role.id)
+        else:
+            member.update(role=None)
```

Comment:
i'm very confused about this case. what are the steps of user action taken to have the omt.role be `None`?

Reply:
I'd hoped the docstring would explain it; 95fb0344c76055bbaa036cb2b43aca10981f30b7 is my second pass at an inline explanation.

The key other side of this interaction is the following method on the `OrganizationMemberTeam` model: 
```python
    def get_team_role(self) -> TeamRole:
        entry_role = roles.get_entry_role(self.organizationmember.role)
        if self.role:
            team_role = team_roles.get(self.role)
            if team_role.priority > entry_role.priority:
                return team_role
        return entry_role
```
Suppose we have an Org Owner, meaning they are entitled to have a minimum team role of Team Admin on every team. Suppose, for whatever reason, they have a bogus `"contributor"` string in their `OrganizationMemberTeam` object. That `if team_role.priority > entry_role.priority` line ensures that we return the Team Admin role to which they're entitled, and correctly ignore the data calling them a Team Contributor.

A lesser problem is when the persistent role is equal to the entry role. Depending on the path they took to get there, we could have some Org Owners and Managers running around with `"admin"` in their `OrganizationMemberTeam` row and others with null. The outcome would be the same but it would be confusing in Redash, etc. There would also be weird corner cases in case one of those users is demoted: a Manager with a persistent `"admin"` string would remain a Team Admin even after being busted down to Org Member, but one with a null string would not. This is bad because the difference would be opaque until it actually happened.

The point of writing null here is essentially: avoid writing an equal or lesser role in the first place so that the database better reflects reality. Except for the "what happens after they're demoted" case, we don't *need* to do this as long as `get_team_role` is checking `if team_role.priority > entry_role.priority`. We also might change our minds, once we have end-to-end demos to play with, whether it's useful to allow team roles that are explicitly lower than entry roles (e.g., someone is an Org Owner but wants to be designated as a Team Contributor, not Admin, because it better reflects their practical responsibilities).

This logic seems to be incurring a bit of a complexity cost, so I'm open to scrapping it in favor of a simpler implementation, if the above either doesn't make sense or doesn't seem to be worth it. If we do, I believe it's important to communicate in the UI design: "here is the team role that you get by virtue of your org role, and here is the team role that you've been manually assigned, be it higher, lower, or equal."